sdk/test/trace/probability_sampler_test.cc

```diff
@@ -0,0 +1,211 @@
+
+	using M = std::map<std::string, int>;
+	M m1 = {{}};
+	opentelemetry::trace::KeyValueIterableView<M> view{m1};
+
+	ProbabilitySampler s1(0.01);
+
+	auto sampling_result = s1.ShouldSample(nullptr, invalid_trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	constexpr uint8_t buf[] = {0, 0, 0, 0, 0, 0, 0, 0x80, 0, 0, 0, 0, 0, 0, 0, 0};
+	opentelemetry::trace::TraceId valid_trace_id(buf);
+
+	sampling_result = s1.ShouldSample(nullptr, valid_trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::NOT_RECORD, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	ProbabilitySampler s2(0.50000001);
+
+	sampling_result = s2.ShouldSample(nullptr, valid_trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	ProbabilitySampler s3(0.49999999);
+
+	sampling_result = s3.ShouldSample(nullptr, valid_trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::NOT_RECORD, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	ProbabilitySampler s4(0.50000000);
+
+	sampling_result = s4.ShouldSample(nullptr, valid_trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+}
+
+TEST(ProbabilitySampler, ShouldSampleWithContext)
+{
+	opentelemetry::trace::TraceId trace_id;
+	opentelemetry::trace::SpanKind span_kind = opentelemetry::trace::SpanKind::kInternal;
+	opentelemetry::sdk::trace::Sampler::SpanContext c1(false, false);
+	opentelemetry::sdk::trace::Sampler::SpanContext c2(true, false);
+	opentelemetry::sdk::trace::Sampler::SpanContext c3(false, true);
+	opentelemetry::sdk::trace::Sampler::SpanContext c4(true, true);
+
+	using M = std::map<std::string, int>;
+	M m1 = {{}};
+	opentelemetry::trace::KeyValueIterableView<M> view{m1};
+
+ 	ProbabilitySampler s1(0.01);
+
+	auto sampling_result = s1.ShouldSample(&c1, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::NOT_RECORD, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	sampling_result = s1.ShouldSample(&c2, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	sampling_result = s1.ShouldSample(&c3, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	sampling_result = s1.ShouldSample(&c4, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+}
+
+TEST(ProbabilitySampler, ProbabilitySamplerHalf)
+{
+	double probability = 0.5;
+	int iterations = 100000, expected_count = iterations * probability,
+		variance = iterations * 0.01;
```

Comment:
Out of curiosity, where is the 0.01 multiplier coming from?

Reply:
It's just an arbitrary value to make sure the tests pass with an acceptable margin or error, in this case I'm giving it a 1% variance